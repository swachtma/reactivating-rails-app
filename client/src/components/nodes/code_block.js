import React, { Component } from 'react';
import { shape, number, string, array, func } from 'prop-types';

import CodeHeader from './code_header';
import CodeSection from './code_section';

const styles = {
  codeBlockStyle: {
    margin: "1.5em 0"
  },
  blockBodyStyle: {
    padding: "10px 0",
    border: "1px solid #021a40",
    backgroundColor: "#F0F0F0",
    overflowX: "scroll"
  }
};

export default class CodeBlock extends Component {
  static propTypes = {
    collapseSections: func,
    attachCopyTrigger: func,
    addSectionRef: func.isRequired,
    node: shape(
      { node_id: number, block_type: string, block_path: string, sections: array }
    ).isRequired
  };

  renderCodeSections = () => {
    let { addSectionRef, node } = this.props;
    return Array.from(node.sections, s => 
      <CodeSection ref={addSectionRef} block_type={node.block_type} 
      priority={s.priority} contents={s.contents} key={s.section_id} section_id={s.section_id} />
    );
  };
  
  render() {
    let { collapseSections, attachCopyTrigger, node } = this.props;
    let { node_id, block_type, block_path } = node;

    return (
      <div style={styles.codeBlockStyle} key={"code_fence_" + node_id}
      id={"code_block_" + node_id} className={"language-" + block_type}>
        { block_path && 
          <CodeHeader block_path={block_path} collapseHandler={collapseSections}
          triggerRef={attachCopyTrigger} /> 
        } 
        <div style={styles.blockBodyStyle}>
          { this.renderCodeSections() }
        </div>
      </div>
    );
  }
}