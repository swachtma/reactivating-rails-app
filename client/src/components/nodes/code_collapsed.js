/* global hljs */
import React, { Component } from 'react'
import { array, string } from 'prop-types'

class CodeCollapsed extends Component {
  static propTypes = {
    section_id: string.isRequired,
    contents: array.isRequired,
    block_type: string.isRequired
  }

  componentDidMount () {
    hljs.highlightBlock(document.getElementById(this.props.section_id + '_collapsed'))
  }

  render () {
    let { contents, section_id, block_type } = this.props
    return (
      <code id={section_id + '_collapsed'}>
        <pre className={block_type}>
          <div>
            {contents[0]}
            <i className="fa fa-arrows-h" aria-hidden="true"></i>
            {contents[contents.length - 1].trim()}
          </div>
        </pre>
      </code>
    )
  }
}

export default CodeCollapsed
