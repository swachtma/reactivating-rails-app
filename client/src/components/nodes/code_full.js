/* global hljs */
import React, { Component } from 'react';

class CodeFull extends Component {
  componentDidMount(){
    hljs.highlightBlock(document.getElementById(this.props.section_id+"_full"));
  }
  
  render() {
    return (
      <code id={this.props.section_id+"_full"}>
        <pre className={this.props.block_type}>{this.props.contents.join("\n")}</pre>
      </code>
    );
  }
}

export default CodeFull;