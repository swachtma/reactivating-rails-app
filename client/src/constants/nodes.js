export const LOAD_NODES = "LOAD_NODES";