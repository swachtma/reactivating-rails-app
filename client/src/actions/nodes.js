import * as constants from '../constants/nodes'

export const loadNodes = (payload) => ({ type: constants.LOAD_NODES, payload })
