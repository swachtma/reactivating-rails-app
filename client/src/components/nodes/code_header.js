import React, { Component } from 'react'
import { Responsive, Popup, Button, Icon } from 'semantic-ui-react'

import * as BREAKPOINTS from '../../constants/breakpoints'

const styles = {
  blockHeaderStyles: {
    borderBottom: '3px double #021a40',
    backgroundColor: '#021a40',
    color: '#fff',
    padding: '5px 10px 5px 20px'
  },
  filePath: {
    fontSize: '.9em',
    display: 'inline-block'
  },
  limitedTitle: {
    display: 'inline-block',
    width: '125px',
    textOverflow: 'ellipsis',
    overflow: 'hidden'
  }
}

class CodeHeader extends Component {
  render () {
    let { block_path, collapseHandler, triggerRef } = this.props
    return (
      <div style={styles.blockHeaderStyles} ref={triggerRef}>
        <FilePath block_path={block_path} />
        <HeaderControlButton handler={() => collapseHandler(false)} icon="plus" label="Expand All" />
        <HeaderControlButton handler={() => collapseHandler(true)} icon="minus" label="Collapse All" />
        <HeaderControlButton icon="copy" label="Copy All" />
      </div>
    )
  }
}

export default CodeHeader

export class FilePath extends Component {
  constructor (props) {
    super(props)
    let path_segments = props.block_path.split('/')
    this.block_file = path_segments[path_segments.length - 1]
    this.block_path = this.props.block_path
  }

  render () {
    let { block_file, block_path } = this
    return (
      <div style={styles.filePath}>
        <Responsive as="span" {...BREAKPOINTS.onlyMobile}><span style={styles.limitedTitle}>{block_file}</span></Responsive>
        <Responsive as="span" {...BREAKPOINTS.aboveMobile}>{block_path}</Responsive>
      </div>
    )
  }
}

export class HeaderControlButton extends Component {
  state = { isOpen: false };

  handleOpen = (e) => {
    this.setState({ isOpen: true })
    if (e.type === 'click') this.handleTimeout()
  }

  handleTimeout = () => {
    this.timeout = setTimeout(() => {
      this.setState({ isOpen: false })
    }, 400)
  }

  handleClose = () => {
    this.setState({ isOpen: false })
    clearTimeout(this.timeout)
  }

  render () {
    let { handler, icon, label } = this.props
    return (
      <span onTouchStart={this.handleTimeout}>
        <Popup
          on={['click', 'hover']}
          open={this.state.isOpen}
          onClose={this.handleClose}
          onOpen={this.handleOpen}
          trigger={
            <Button className={icon} onClick={handler}
              icon size="mini" floated="right" color="grey">
              <Icon name={icon} />
            </Button>
          }
          position="bottom center" content={label}
        />
      </span>
    )
  }
}
