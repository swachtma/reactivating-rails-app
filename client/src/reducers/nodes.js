import * as constants from '../constants/nodes';

let default_state = [];

export const sortNodes = (nodes) => {
  return nodes.sort((a,b) =>{
    if(a.chapter_id === b.chapter_id){
      return a.id - b.id;
    } else {
      return a.chapter_id - b.chapter_id;
    }
  });
};

const nodesReducer = function(state = default_state,action){
  switch (action.type) {
    case(constants.LOAD_NODES):
      return [...sortNodes(action.payload)];
    default:
      return state;
  }
};

export default nodesReducer;