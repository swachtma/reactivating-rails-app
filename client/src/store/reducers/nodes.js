import * as constants from '../../constants/nodes'

let default_state = []

export const sortNodes = (nodes) =>
  nodes.sort((a, b) => a.chapter_id === b.chapter_id ? a.id - b.id : a.chapter_id - b.chapter_id)
  // 2 step sort, compares by node id unless chapters IDs not matched, then sort by chapter ID

export default function nodesReducer (state = default_state, action) {
  switch (action.type) {
    case (constants.LOAD_NODES):
      return [...sortNodes(action.payload)]
    default:
      return state
  }
}
