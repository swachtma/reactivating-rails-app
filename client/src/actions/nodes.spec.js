/* global expect */
import * as actions from './nodes'

describe('Nodes actionCreators', () => {
  test('loadNodes(node)', () => {
    expect(actions.loadNodes([{ contents: 'Test node' }])).toMatchSnapshot()
  })
})
