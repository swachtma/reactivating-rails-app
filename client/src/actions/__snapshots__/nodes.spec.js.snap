// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nodes actionCreators loadNodes(node) 1`] = `
Object {
  "payload": Array [
    Object {
      "contents": "Test node",
    },
  ],
  "type": "LOAD_NODES",
}
`;
